bpl: bpl_cfg: move stop_on_failure_attempts from BCL to BPL

mariomaz opened the pull request:
Parameter stop_on_failure_attempts is located in BCL configuration so
it cannot be easily changed during the certification setup. Moving it
to BPL configuration allows it to be changed with an init script when
in certification mode.

Create bpl_cfg_get_stop_on_failure_attempts(). Remove parameter
stop_on_failure_attempts from configuration files. Use the new function
where required.

Create bpl_cfg_get_certification_mode(). Remove parameter
certification_mode from configuration files. Propagate the parameter
value from agent to controller when slave joins.

Modify prplmesh_utils.sh to add options to enable / disable
certification mode.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

arnout commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -288,12 +286,11 @@ static int run_beerocks_slave(beerocks::config_file::sConfigSlave &beerocks_slav
-    son::slave_thread::sSlaveConfig son_slave_conf;
-    fill_son_slave_config(beerocks_slave_conf, son_slave_conf, 0);
+    // read the number of failures allowed before stopping agent from platform configuration
+    int stop_on_failure_attempts = bpl_cfg_get_stop_on_failure_attempts();
```
For the time being, only the platform manager thread should make bpl calls. That should be changed because we want e.g. controller to also be able to access bpl - cfr. #542. 

As it happens, though, the platform manager doesn't run in a separate thread (even though it's called main_thread, the thread is never start()-ed). Therefore, we *can* access bpl from this function. Only, we have to make sure it's done after bpl_init(). Therefore, the entire code flow below should be changed:

- platform_mgr.init()
- read stop_on_failure_attempts
- construct and start backhaul_mgr
- run the `g_running` loop

mariomaz replied:
Done

arnout commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -391,6 +388,9 @@ static int run_son_slave(int slave_num, beerocks::config_file::sConfigSlave &bee
     fill_son_slave_config(beerocks_slave_conf, son_slave_conf, slave_num);
 
+    // read the number of failures allowed before stopping agent from platform configuration
+    son_slave_conf.stop_on_failure_attempts = bpl_cfg_get_stop_on_failure_attempts();
```
Again, it's not possible to access bpl here. We're in a different process even (this is the `beerocks_agent -i <iface>` process, not the plain agent process).

So I think we have to initialize stop_on_failure_attempts to 0 and get it from platform_manager as part of cACTION_PLATFORM_SON_SLAVE_REGISTER_RESPONSE instead.

mariomaz replied:
Done. You are probably aware that if a timeout occurs while in STATE_WAIT_FOR_PLATFORM_MANAGER_REGISTER_RESPONSE, then thread will never stop but retry indefinitely because of that initial value that never gets updated.

arnout reviewed (commented):
I don't think it will work like this - probably that's why it's still a draft PR.

It would be better to split it into three commits:

1. add stop_on_failure_attempts and certification_mode to bpl
2. use stop_on_failure_attempts from bpl instead of config
3. use certification_mode from bpl instead of from config

arnout commented:
> Done, if I've understood you correctly.
> I've realized that bpl_get_xxx() functions return -1 when parameter is not found. IMHO it would be more appropriate that functions returned a default value in such case instead of an error. Moreover, since return value is an int with either value read or error code, it's not possible to distinguish between them if -1 were a valid value for some parameter.

I completely agree with this. I created #568 (which covers quite a bit more) for this.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -1023,6 +1023,7 @@ bool slave_thread::handle_cmdu_control_message(
             return false;
         }
         configuration_stop_on_failure_attempts = request_in->attempts();
+        stop_on_failure_attempts               = configuration_stop_on_failure_attempts;
```
Why do we need both `configuration_stop_on_failure_attempts` and `stop_on_failure_attempts`?

arnout replied:
Because stop_on_failure_attempts is decremented on every failure, but reset to its original value on success. E.g. if you set stop_on_failure_attempts to 5, and you have a failure every hour or so but successes between those failures, you don't want it to stop after 5 hours.

That said, I don't think this is very useful. Either you stop on failure, or you retry forever.

mariomaz committed:
bpl: bpl_cfg: add stop_on_failure_attempts and certification_mode to BPL

Parameter stop_on_failure_attempts is located in BCL configuration so
it cannot be easily changed during the certification setup. Moving it
to BPL configuration allows it to be changed with an init script when
in certification mode.

Prepare for "bpl: use stop_on_failure_attempts from BPL instead of
from BCL" and "bpl: use certification_mode from BPL instead of from
BCL".

Create bpl_cfg_get_stop_on_failure_attempts() and
bpl_cfg_get_certification_mode() to read parameters from BPL.

Modify platform_manager_thread to use the new functions to read
parameters and add them to platform settings.

Modify prplmesh_utils.sh to enable certification mode in configuration.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
bpl: use stop_on_failure_attempts from BPL instead of BCL

In backhaul manager, use the new function
bpl_cfg_get_stop_on_failure_attempts() to read parameter from BCL
instead of BCL.

In slave_thread, initialize stop_on_failure_attempts to 0 and get it
from platform_manager as part of
cACTION_PLATFORM_SON_SLAVE_REGISTER_RESPONSE instead

Remove parameter stop_on_failure_attempts from configuration files.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

mariomaz committed:
bpl: use certification_mode from BPL instead of from BCL

Use the new function bpl_cfg_get_certification_mode() to read parameter
from BCL instead of BCL. Propagate the parameter value from agent to
controller when slave joins.

Remove parameter certification_mode from configuration files.

Signed-off-by: Mario Maz <mmazzapater@gmail.com>

arnout reviewed (approved):
Excellent work!

tomereli commented on common/beerocks/scripts/prplmesh_utils.sh.in:
```diff
@@ -84,11 +84,17 @@ platform_deinit() {
     operating_mode=${2-Gateway}
+    
+    # In certification, agent must retry onboarding indefinitely
+    certification_mode="1"
```
Defining these variables and setting them right after is redundant, just do
```
echo "certification_mode=1"
echo "stop_on_failure_attempts=0"
```
(nitpick)

mariomaz replied:
Done

tomereli commented on common/beerocks/bcl/include/bcl/beerocks_config_file.h:
```diff
@@ -51,7 +51,6 @@ class config_file {
         std::string load_health_check;
```
commit message typo - "from BCL instead of BCL." --> "from BPL instead of BCL"

tomereli reviewed (approved):
small comments - but I agree with @arnout , excellent work.
LGTM.

arnout commented on agent/src/beerocks/slave/beerocks_slave_main.cpp:
```diff
@@ -391,6 +389,10 @@ static int run_son_slave(int slave_num, beerocks::config_file::sConfigSlave &bee
     son::slave_thread::sSlaveConfig son_slave_conf;
     fill_son_slave_config(beerocks_slave_conf, son_slave_conf, slave_num);
 
+    // disable stopping on failure initially. Later on, it will be read from BPL as part of
+    // cACTION_PLATFORM_SON_SLAVE_REGISTER_RESPONSE
+    son_slave_conf.stop_on_failure_attempts = 0;
```
Since Tomer was anyway giving nitpick comments, I have one too: I think it would be nicer if this were in fill_son_slave_config() instead.

mariomaz replied:
Done

tomereli reviewed (approved):
LGTM. Please rebase and merge.
